### PR TITLE
[MON-1695] expose /api/v1/labels end point for Thanos query.

### PR DESCRIPTION
**Pull request by raptorsun**
* [x] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

Expose /api/v1/labels end point for Thanos query.
End point for /api/v1/label/<name>/values is added but not working yet, need to update kube-rbac-proxy to support at least trailing wildcard patterns. 

This PR requires the 2 PRs below done before merge, and they are merged now:
1. [This PR](https://github.com/brancz/kube-rbac-proxy/pull/135) is merged for allowing trailing wildcard in kube-rbac-proxy  --allow-paths .
2. [This PR](https://github.com/openshift/kube-rbac-proxy/pull/50) is merged for version bump to 0.11.

### CODE REVIEW

**Review by PhilipGough — APPROVED** [review]
/lgtm

**Comment by PhilipGough** [comment]
/hold

do we want to wait for the kube-rbac-proxy change before testing/merging?

**Comment by raptorsun** [comment]
> /hold
> 
> do we want to wait for the kube-rbac-proxy change before testing/merging?

yes, let's hold it before modifying kube-rbac-proxy.

**Comment by raptorsun** [comment]
[This PR](https://github.com/brancz/kube-rbac-proxy/pull/135) is created for allowing trailing wildcard in kube-rbac-proxy --allow-paths .

**Comment by raptorsun** [comment]
now we are waiting for a new release from kube-rbac-proxy that includes [This PR](https://github.com/brancz/kube-rbac-proxy/pull/135)

File: jsonnet/thanos-querier.libsonnet [hunk 1]
```diff
@@ -493,6 +493,7 @@ function(params)
                   '--insecure-listen-address=127.0.0.1:9095',
                   '--upstream=http://127.0.0.1:9090',
                   '--label=namespace',
+                  '-enable-label-apis',
```
Comment (paulfantom) [inline jsonnet/thanos-querier.libsonnet:496]:
```suggestion
                  '--enable-label-apis',
```

Comment (raptorsun) [reply jsonnet/thanos-querier.libsonnet]: updated, thanks 👍

**Comment by raptorsun** [comment]
[This required PR](https://github.com/openshift/kube-rbac-proxy/pull/50) is waiting for Openshift 4.10 development branch to open on  03SEP2021.

**Comment by paulfantom** [comment]
@raptorsun https://github.com/openshift/kube-rbac-proxy/pull/50 was merged. This PR is clear to proceed.

**Comment by raptorsun** [comment]
@paulfantom Thank you very much, this is much sooner than expected! 🚀

**Comment by simonpasquier** [comment]
/hold
waiting for 4.10 development to open

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,8 @@
 # Note: This CHANGELOG is only for the monitoring team to track all monitoring related changes. Please see OpenShift release notes for official changes.
 
+## 4.10
+
+- [#1299](https://github.com/openshift/cluster-monitoring-operator/pull/1299) Expose expose /api/v1/labels endpoint for Thanos query.
```
Comment (simonpasquier) [inline CHANGELOG.md:5]:
```suggestion
- [#1299](https://github.com/openshift/cluster-monitoring-operator/pull/1299) Expose expose /api/v1/labels endpoint for Thanos query.

```

**Comment by simonpasquier** [comment]
Can you add an e2e test to verify that an authenticated user can access the endpoints (similar to `assertTenancyForMetrics()` and `assertTenancyForRules()`?

**Comment by raptorsun** [comment]
Good idea, new e2e test will be added for the new endpoints.

File: test/e2e/user_workload_monitoring_test.go [hunk 3]
```diff
@@ -860,6 +864,141 @@ func assertTenancyForRules(t *testing.T) {
...
 
+func assertTenancyForLabels(t *testing.T) {
+	const testAccount = "test-labels"
+
+	_, err := f.CreateServiceAccount(userWorkloadTestNs, testAccount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Grant enough permissions to read labels. todo: check if need create new role
```
Comment (simonpasquier) [inline test/e2e/user_workload_monitoring_test.go:875]: The labels endpoints are similar to the query endpoints: if you are authorized for /api/v1/query, you should also be for /api/v1/labels.

File: test/e2e/user_workload_monitoring_test.go [hunk 4]
```diff
@@ -860,6 +864,141 @@ func assertTenancyForRules(t *testing.T) {
...
+		token,
+		&framework.QueryParameterInjector{
+			Name:  "namespace",
+			Value: userWorkloadTestNs,
+		},
+	)
+
+	t.Logf("Checking all labels")
+
+	// check /api/v1/labels end point
```
Comment (simonpasquier) [inline test/e2e/user_workload_monitoring_test.go:912]:
```suggestion
	// check /api/v1/labels endpoint.
```

File: test/e2e/user_workload_monitoring_test.go [hunk 5]
```diff
@@ -860,6 +864,141 @@ func assertTenancyForRules(t *testing.T) {
...
+
+		return nil
+
+	})
+	if err != nil {
+		t.Fatalf("failed to query labels from Thanos querier: %v", err)
+	}
+
+	// check /api/v1/label/<label>/value using some examples
+	for _, label := range []string{"endpoint", "event", "node"} {
```
Comment (simonpasquier) [inline test/e2e/user_workload_monitoring_test.go:951]: I would check that the `namespace` label returns only 1 item (e.g. `userWorkloadTestNs`).

**Comment by juzhao** [comment]
can get /api/v1/labels and /api/v1/label/$({label_name}/values for user project
```
# curl -k -H "Authorization: Bearer $token" 'https://thanos-querier.openshift-monitoring.svc:9092/api/v1/labels?namespace=ns1' | jq
{
  "status": "success",
  "data": [
    "__name__",
    "alertname",
    "alertstate",
    "cluster_ip",
    "condition",
    "configmap",
    "container",
    "container_id",
    "cpu",
    "created_by_kind",
    "created_by_name",
    "deployment",
    "device",
    "endpoint",
    "host_ip",
    "host_network",
    "id",
...
}

# curl -k -H "Authorization: Bearer $token" 'https://thanos-querier.openshift-monitoring.svc:9092/api/v1/label/alertname/values?namespace=ns1'  | jq
{
  "status": "success",
  "data": [
    "TestAlert"
  ]
}
```
/label qe-approved

File: test/e2e/user_workload_monitoring_test.go (hunk 5 above)
Comment (raptorsun) [reply test/e2e/user_workload_monitoring_test.go]: The test has been updated to include the namespace label.

File: CHANGELOG.md (hunk 2 above)
Comment (simonpasquier) [reply CHANGELOG.md:5]:
```suggestion
- [#1299](https://github.com/openshift/cluster-monitoring-operator/pull/1299) Expose /api/v1/labels endpoint for Thanos query.
```

File: test/e2e/user_workload_monitoring_test.go (hunk 3 above)
Comment (simonpasquier) [reply test/e2e/user_workload_monitoring_test.go:875]: can you remove the `todo: ...` part (or address it)?

File: test/e2e/user_workload_monitoring_test.go (hunk 4 above)
Comment (simonpasquier) [reply test/e2e/user_workload_monitoring_test.go:912]: still need fixing :)

File: test/e2e/user_workload_monitoring_test.go [hunk 6]
```diff
@@ -860,6 +864,193 @@ func assertTenancyForRules(t *testing.T) {
...
+		}
+
+		return nil
+
+	})
+	if err != nil {
+		t.Fatalf("failed to query labels from Thanos querier: %v", err)
+	}
+
+	// check /api/v1/label/<label>/value using some examples
```
Comment (simonpasquier) [inline test/e2e/user_workload_monitoring_test.go:954]: I think that checking `/api/v1/label/namespace/values` is enough

**Review by simonpasquier — COMMENTED** [review]
/hold

still some comments to be addressed.

File: test/e2e/user_workload_monitoring_test.go (hunk 3 above)
Comment (raptorsun) [reply test/e2e/user_workload_monitoring_test.go:875]: yup. already checked: no need for another account

File: test/e2e/user_workload_monitoring_test.go (hunk 6 above)
Comment (raptorsun) [reply test/e2e/user_workload_monitoring_test.go]: tests for other labels have been removed.

**Comment by raptorsun** [comment]
> /hold
> 
> still some comments to be addressed.

E2E test has been updated according to comments.

**Pull request merged**